### PR TITLE
Distributed searchlight

### PR DESCRIPTION
**Pull request by mjanderson09**

### CODE REVIEW

**Comment by yidawang** [comment]
Please provide more comments to describe at least the public methods. I found it a bit confused when reading the code now. For example, I am lost when trying to understand how the ownership is distributed.

File: examples/searchlight/example_searchlight.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+import numpy as np
+from mpi4py import MPI
+import sys
+
+from brainiak.searchlight import example_fn
+from brainiak.searchlight.searchlight import Searchlight
+
+"""Distributed Searchlight Example
+"""
```
Comment (mihaic) [inline examples/searchlight/example_searchlight.py]: Document the way this code supposed is supposed to be executed.

File: brainiak/searchlight/searchlight.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                   searchlight cube
+
+        max_blk_edge: max edge length, in voxels, of the 3D block
+
+        """
+        self.sl_rad = sl_rad
+        self.max_blk_edge = max_blk_edge
+
+    def _get_ownership(self, data):
+        """Determine which processes own each subject
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: "process owns"?

File: brainiak/searchlight/searchlight.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+          - a triple containing the top left point of the block and
+          - a triple containing the size in voxels of the block
+
+        Returns
+        -------
+
+        In the case of a 3D array, a 3D subarray at the block location
+
+        In the case of a 4D array, a 4D subarray at the block location,
+        including the entire 0th dimension.
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: Isn't time the last dimension?

File: brainiak/searchlight/searchlight.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    larger portions of the volume called blocks.
+    """
+    def __init__(self, sl_rad=1, max_blk_edge=10):
+        """Constructor
+
+        Parameters
+        ----------
+
+        sl_rad: radius, in voxels, of the sphere inscribed in the
+                   searchlight cube
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: Does this include the center or not?

File: brainiak/searchlight/searchlight.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+    def distribute(self, subj, mask):
+        """Distribute data to processes
+
+        Parameters
+        ----------
+
+        subj: list of 4D arrays containing data for one or more subjects.
+              Each entry of the list must contaion a 4D array on only one
+              MPI rank, and "None" on all other ranks.
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: Aren't we allowing all data to be loaded on a single rank?

File: brainiak/searchlight/searchlight.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+        mask: 3D array with nonzero values at active vertices
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: s/nonzero/true

File: brainiak/searchlight/searchlight.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+        """
+        self.mask = mask
+        comm = MPI.COMM_WORLD
+        rank = comm.rank
+
+        # Get/set ownership
+        ownership = self._get_ownership(subj)
+        full_pts = self._get_blocks(mask) if rank == 0 else None
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: What is a point?

File: brainiak/searchlight/searchlight.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        """Distribute data to processes
+
+        Parameters
+        ----------
+
+        bcast_var:    shared data which is broadcast to all processes
+
+        """
+
+        comm = MPI.COMM_WORLD
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: How about making `comm` an instance attribute?

File: brainiak/searchlight/searchlight.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+        """Perform a function for each block in a volume.
+
+        Parameters
+        ----------
+
+        block_fn: function to apply to each block:
+
+                Parameters
+
+                subj: list of 4D arrays containing subset of subject data
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: s/subj/subjs?

Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: The padding should be documented here.

File: brainiak/searchlight/searchlight.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+                    for i in range(0, mat.shape[0]):
+                        for j in range(0, mat.shape[1]):
+                            for k in range(0, mat.shape[2]):
+                                outmat[pt[0]+self.sl_rad+i,
+                                       pt[1]+self.sl_rad+j,
+                                       pt[2]+self.sl_rad+k] = mat[i, j, k]
+
+        return outmat
+
+    def searchlight_voxel(self, voxel_fn, pool_size=None):
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: How about `run_searchlight`? And `run_block_function` for `searchlight_block`?

File: brainiak/searchlight/searchlight.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+        all_blocks = self.comm.bcast(all_blocks)
+
+        # Divide data and mask
+        splitsubj = [self._split_volume(s, all_blocks)
+                     if s is not None else None
+                     for s in subj]
+        submasks = self._split_volume(mask, all_blocks)
+
+        # Scatter points, data, and mask
+        self.pts = self._scatter_list(all_blocks, 0)
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: What is a point? See `pt` in the example as well.

File: brainiak/searchlight/searchlight.py [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    def _get_block_data(self, mat, block):
+        """Retrieve a block from a 3D or 4D volume
+
+        Parameters
+        ----------
+
+        mat: a 3D or 4D volume
+
+        block: a tuple containing block information:
+          - a triple containing the top left point of the block and
```
Comment (mihaic) [inline brainiak/searchlight/searchlight.py]: Top left front or top left back? Also, wouldn't it be better to remove direction altogether and say "the voxel with the lowest coordinates"?

**Pull request merged**